xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -373,33 +409,64 @@ private Set<String> getResourceKeys(XdsResourceType<?> xdsResourceType) {
     return resourceSubscribers.get(xdsResourceType).keySet();
   }
 
-  private void cleanUpResourceTimers() {
+  // cpcForThisStream is null when doing shutdown
+  private void cleanUpResourceTimers(ControlPlaneClient cpcForThisStream) {
+    List<String> authoritiesForCpc = getActiveAuthorities(cpcForThisStream);
+
     for (Map<String, ResourceSubscriber<?>> subscriberMap : resourceSubscribers.values()) {
       for (ResourceSubscriber<?> subscriber : subscriberMap.values()) {
-        subscriber.stopTimer();
+        if (cpcForThisStream == null || authoritiesForCpc.contains(subscriber.authority)) {
+          subscriber.stopTimer();
+        }
+      }
+    }
+  }
+
+  public ControlPlaneClient getOrCreateControlPlaneClient(ImmutableList<ServerInfo> serverInfos) {
+    if (serverInfos == null || serverInfos.isEmpty()) {
+      return null;
+    }
+
+    for (ServerInfo serverInfo : serverInfos) {
+      if (serverCpClientMap.containsKey(serverInfo)) {
+        ControlPlaneClient controlPlaneClient = serverCpClientMap.get(serverInfo);
+        if (controlPlaneClient.isInBackoff()) {
+          continue;
+        }
+        return controlPlaneClient;
+      } else {
+        ControlPlaneClient cpc = getOrCreateControlPlaneClient(serverInfo);
```

Comment:
Use this method unconditionally? The log statement here isn't even doing anything, as you also added a log statement to `getOrCreateControlPlaneClient(ServerInfo)`

```java
    for (ServerInfo serverInfo : serverInfos) {
      ControlPlaneClient cpc = getOrCreateControlPlaneClient(serverInfo);
      if (cpc.isInBackoff()) {
        continue;
      }
      return cpc;
    }
```

You don't even need the `isEmpty()` check.

Reply:
done.